Refactor types and add alignment to them

nurmukhametov opened the pull request:
This PR adds alignment field to Type. This is needed to implement #2145

aneshlya commented on src/type.h:
```diff
@@ -1169,9 +1141,16 @@ class FunctionType : public Type {
+    FunctionType(const FunctionType &other);
+
+    const FunctionType *Clone() const override;
+    const FunctionType *CloneWithRetTypeAndNewParamTypes(const Type *newReturnType,
```
Maybe rename it to `CloneWithSignature` but it's up to you.

nurmukhametov replied:
Done.

aneshlya reviewed (commented):
Thanks for this huge refactoring!

It mostly looks good! I'm slightly concerned by the extensive usage of `mutable` which should be carefully checked and documented.

I don't really like the `Clone...` functions since we now have a mix of `GetAs...` and `Clone...` which is confusing. Since these are private methods, I think we should use lower camel case to visually distinguish them. Also, using `createWith..`. or `deriveWith...` seems more appropriate here since we're creating new objects with modifications rather than making deep copies.

Consider standardizing on one consistent naming pattern for all these type transformation methods to improve readability and maintainability going forward.

nurmukhametov commented on src/parse.yy:
```diff
@@ -1410,18 +1410,18 @@ type_specifier_list
-    | TOKEN_DOUBLE { $$ = AtomicType::UniformDouble->GetAsUnboundVariabilityType(); }
-    | TOKEN_INT64 { $$ = AtomicType::UniformInt64->GetAsUnboundVariabilityType(); }
-    | TOKEN_UINT64 { $$ = AtomicType::UniformUInt64->GetAsUnboundVariabilityType(); }
+    | TOKEN_BOOL    { $$ = CastType<AtomicType>(AtomicType::UniformBool->GetAsUnboundVariabilityType()); }
```
This is needed because `GetAsUnboundVariabilityType` is no longer overridden in `AtomicType` and now returns Type (not `AtomicType` as before). By the way, having different semantic types like `atomicType` for `atomic_var_type_specifier`, `int_constant_type`, and `template_int_constant_type` doesn't make much sense because the parser has no places that use different APIs (actually, only the field `basicType` is an extra interface of `AtomicType`). I have removed `atomicType`.

nurmukhametov commented on src/type.cpp:
```diff
@@ -2145,6 +1987,30 @@ StructType::StructType(const std::string &n, const llvm::SmallVector<const Type
+}
+
+const StructType *StructType::CloneWithAlignment(unsigned int newAlignment) const {
+    return new StructType(name, elementTypes, elementNames, elementPositions, isConst, variability, isAnonymous, pos,
```
It is different from the `Type::createWithAlignment` because there there are two steps:
```c++
auto *s = new StructType(..., alignment); // create with the old alignment
s->alignment = newAlignment; // set the new alignment
```
Since, `StructType` constructor performs non-trivial operations, e.g. creates `m->structTypeMap` entry, that depend on the arguments of the constructor, we need to call the constructor with the new alignment.

nurmukhametov commented on src/type.h:
```diff
@@ -1096,69 +1061,76 @@ class FunctionType : public Type {
     std::string mangleTemplateArgs(TemplateArgs *templateArgs) const;
 
-    const Type *const returnType;
+    mutable const Type *returnType;
```
It was needed only to change the values of fields via `const Type*` (all `createWith...` do this).
I changed `create` to return `Type *` (not `const Type*`), which allowed me to remove all mutables. Unfortunately, I had to remove const from the fields that are modified in the `createWith...` methods after the object has been created.

nurmukhametov commented on src/type.h:
```diff
@@ -1096,69 +1061,76 @@ class FunctionType : public Type {
+    bool IsSafe() const { return flags & FUNC_SAFE; }
+
+    /** Set the 'safe' flag for the function. */
+    void SetSafe(bool safe) {
```
I don't understand why it is needed. We had only one place where this function was used. I refactored it by passing the proper flags and `costOverride` to the constructor. This allowed me to make `costOverride` private as well.

aneshlya replied:
> I don't understand why it is needed.

For completeness of the interface. The question which arises is why do we have special function to set safe flag and not other flags? Anyway it was optional suggestion, if you don't want to do it, that's OK :)

nurmukhametov commented:
> It mostly looks good! I'm slightly concerned by the extensive usage of `mutable` which should be carefully checked and documented.

I have removed `mutable`; see the detailed comments in the other comment.

> I don't really like the `Clone...` functions since we now have a mix of `GetAs...` and `Clone...` which is confusing. Since these are private methods, I think we should use lower camel case to visually distinguish them. Also, using `createWith..`. or `deriveWith...` seems more appropriate here since we're creating new objects with modifications rather than making deep copies.

I have renamed them to `create...`

aneshlya reviewed (approved):
LGTM!
